src/fairseq2/nn/lora.py

```diff
@@ -0,0 +1,373 @@
+            nn.init.zeros_(self.lora_B)
+
+    @property
+    def wrapped_module(self) -> nn.Module:
+        return self
+
+    def merge(self) -> None:
+        if self.merged:
+            return
+        self.weight.data += self.lora_B.data @ self.lora_A.data * self.scaling
+
+        self.merged = True
+
+    def unmerge(self) -> None:
+        if not self.merged:
+            return
+        else:
+            self.weight.data -= self.lora_B.data @ self.lora_A.data * self.scaling
+
+            self.merged = False
+
+
+def wrap_lora(
+    module: nn.Module, config: LoRAConfig, skip_init: bool = False
+) -> nn.Module:
+    """Iterate all submodules. If `config.keys` regex matches module name,
+    wrap it with a LoRALayer based on module's type. Note that the wrapping
+    also happens in-place."""
+    for name, submodule in module.named_modules():
+        if not _is_target_module(name, config.keys):
+            continue
+
+        submodule_path = name.split(".")
+        parent = module.get_submodule(".".join(submodule_path[:-1]))
+        submodule_name = submodule_path[-1]
+
+        lora_layer: Optional[LoRALayer] = None
+        if isinstance(submodule, Projection):
+            lora_layer = LoRALinear(
+                wrapped=submodule,
+                config=config,
+                skip_init=skip_init,
+                device=submodule.weight.device,  # type: ignore[arg-type]
+                dtype=submodule.weight.dtype,  # type: ignore[arg-type]
+            )
+        elif isinstance(submodule, Embedding):
+            lora_layer = LoRAEmbedding(
+                wrapped=submodule,
+                config=config,
+                device=submodule.weight.device,  # type: ignore[arg-type]
+                dtype=submodule.weight.dtype,  # type: ignore[arg-type]
+            )
+        else:
+            raise ValueError(
+                f"Cannot wrap the module '{name}' with LoRA as the module type '{type(submodule).__name__}' is not supported."
+            )
+
+        lora_layer.train(mode=submodule.training)
+        setattr(parent, submodule_name, lora_layer)
+
+    return module
+
+
+def unwrap_lora(module: nn.Module, merge: bool = True) -> nn.Module:
+    # Reverses the model to its original architecture by replacing all
+    # `LoRALayers` back with their original wrapped modules.
+    # By default, we perform `merge_lora` before unwrapping.
+    # Note that the unwrapping also happends in-place.
+    if merge:
+        merge_lora(module)
+
+    for name, submodule in module.named_modules():
+        if not isinstance(submodule, LoRALayer):
+            continue
+
+        submodule_path = name.split(".")
+        parent = module.get_submodule(".".join(submodule_path[:-1]))
+        submodule_name = submodule_path[-1]
+
+        unwrapped_layer: Optional[nn.Module] = None
+        if isinstance(submodule, LoRALinear):
+            # TODO: currently there's no way to distinguish which type the
```

Comment:
Please see my comment above regarding this.

Reply:
Updated in the new version